api/api_ethereum.go

```diff
@@ -917,3 +1166,164 @@ func (api *EthereumAPI) rpcMarshalHeader(head *types.Header) (map[string]interfa
+	}
+	return res, gas, nil
+}
+
+func EthDoEstimateGas(ctx context.Context, b Backend, args EthTransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, gasCap uint64) (hexutil.Uint64, error) {
+	// Binary search the gas requirement, as it may be higher than the amount used
+	var (
+		lo  uint64 = params.TxGas - 1
+		hi  uint64
+		cap uint64
+	)
+	// Use zero address if sender unspecified.
+	if args.From == nil {
+		args.From = new(common.Address)
+	}
+	// Determine the highest gas limit can be used during the estimation.
+	if args.Gas != nil && uint64(*args.Gas) >= params.TxGas {
+		hi = uint64(*args.Gas)
+	} else {
+		// Ethereum set hi as gas ceiling of the block but,
+		// there is no actual gas limit in Klaytn, so we set it as params.UpperGasLimit.
+		hi = params.UpperGasLimit
+	}
+	// Normalize the max fee per gas the call is willing to spend.
+	var feeCap *big.Int
+	if args.GasPrice != nil && (args.MaxFeePerGas != nil || args.MaxPriorityFeePerGas != nil) {
+		return 0, errors.New("both gasPrice and (maxFeePerGas or maxPriorityFeePerGas) specified")
+	} else if args.GasPrice != nil {
+		feeCap = args.GasPrice.ToInt()
+	} else if args.MaxFeePerGas != nil {
+		feeCap = args.MaxFeePerGas.ToInt()
+	} else {
+		feeCap = common.Big0
+	}
+	// recap the highest gas limit with account's available balance.
+	if feeCap.BitLen() != 0 {
+		state, _, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+		if err != nil {
+			return 0, err
+		}
+		balance := state.GetBalance(*args.From) // from can't be nil
+		available := new(big.Int).Set(balance)
+		if args.Value != nil {
+			if args.Value.ToInt().Cmp(available) >= 0 {
+				return 0, errors.New("insufficient funds for transfer")
+			}
+			available.Sub(available, args.Value.ToInt())
+		}
+		allowance := new(big.Int).Div(available, feeCap)
+
+		// If the allowance is largser than maximum uint64, skip checking
+		if allowance.IsUint64() && hi > allowance.Uint64() {
+			transfer := args.Value
+			if transfer == nil {
+				transfer = new(hexutil.Big)
+			}
+			logger.Warn("Gas estimation capped by limited funds", "original", hi, "balance", balance,
+				"sent", transfer.ToInt(), "maxFeePerGas", feeCap, "fundable", allowance)
+			hi = allowance.Uint64()
+		}
+	}
+	// Recap the highest gas allowance with specified gascap.
+	if gasCap != 0 && hi > gasCap {
+		logger.Warn("Caller gas above allowance, capping", "requested", hi, "cap", gasCap)
+		hi = gasCap
+	}
+	cap = hi
+
+	// Create a helper to check if a gas allowance results in an executable transaction
+	executable := func(gas uint64) (bool, error) {
+		args.Gas = (*hexutil.Uint64)(&gas)
+		_, _, err := EthDoCall(ctx, b, args, rpc.NewBlockNumberOrHashWithNumber(rpc.LatestBlockNumber), nil, 0, gasCap)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	// Execute the binary search and hone in on an executable gas limit
+	for lo+1 < hi {
+		mid := (hi + lo) / 2
+		isExecutable, _ := executable(mid)
```

Comment:
We don't need to check err here?
https://github.com/ethereum/go-ethereum/blob/eae3b1946a276ac099e0018fc792d9e8c3bfda6d/internal/ethapi/api.go#L1077

Reply:
To handle error just like Ethereum logic, I updated some codes in [this commit](https://github.com/klaytn/klaytn/pull/1134/commits/0c0f6c8792c4b9a8b4ee62a5e3538dad36870ff3).

* Now we can specify whether the error is occurred by EVM execution or not.

After that commit,

**KEN: Test with revert situation(Call function required owner from address which is not owner)**
```js
> eth.estimateGas({from: "0x3753f5bf8cCA929EeeB4B74F200dfc6375ad5444", to: "0xbe3892d33620be5aca8c75d39e7401871194d290", data: "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"})


Error: execution reverted: Caller is not owner
    at web3.js:3278:20
    at web3.js:6805:15
    at web3.js:5216:36
    at <anonymous>:1:1
```

**Geth: Test with revert situation((Call function required owner from address which is not owner)**
```js
> eth.estimateGas({from: "0xca7a99380131e6c76cfa622396347107aeedca2d", to: "0x27a110C35BA478034d7F874C89787E0FE587f65F", data: "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"})
Error: execution reverted: Caller is not owner
	at web3.js:6357:37(47)
	at web3.js:5091:62(37)
	at <eval>:1:16(10)
```


FYI. In klay namespace, it cannot specify the error is occurred by VM or not.
> klay.estimateGas({from: "0x3753f5bf8cCA929EeeB4B74F200dfc6375ad5444", to: "0xbe3892d33620be5aca8c75d39e7401871194d290", data: "0xa6f9dae1000000000000000000000000ca7a99380131e6c76cfa622396347107aeedca2d"})


Error: gas required exceeds allowance or always failing transaction
    at web3.js:3278:20
    at web3.js:6805:15
    at web3.js:5216:36
    at <anonymous>:1:1

Reply:
Ohhh... looks good! This might be good to applied in klay in the future. Thank you